lib/AFP/AbstractFilePath.hs

```diff
@@ -0,0 +1,723 @@
+--    mentions that it requires a UTF8 compatible system. These things should be documented.
+-- 3. When dealing with user input (e.g. on the command line) on /unix/ as e.g. @String@ the input
+--    encoding is lost. The output encoding (e.g. how we write a filename to disk) can then
+--    either follow the current locale again ('toAbstractFilePath'') or a fixed encoding
+--    ('toAbstractFilePath'). The decision should be clearly documented. If the input is in the
+--    form of a @ByteString@, then 'bsToAFP' may be of interest, unless the input needs further
+--    interpretation.
+
+module AFP.AbstractFilePath
+  (
+  -- * Types
+    AbstractFilePath
+  , OsString
+  , OsWord
+
+  -- * Construction
+  , toAbstractFilePath
+  , toAbstractFilePathIO
+  , bsToAFP
+  , afp
+  , packAFP
+
+  -- * Deconstruction
+  , fromAbstractFilePath
+  , fromAbstractFilePathIO
+  , unpackAFP
+
+  -- * Word construction
+  , fromChar
+
+  -- * Separator predicates
+  , pathSeparator
+  , pathSeparators
+  , isPathSeparator
+  , searchPathSeparator
+  , isSearchPathSeparator
+  , extSeparator
+  , isExtSeparator
+
+  -- * $PATH methods
+  , splitSearchPath
+
+  -- * Extension functions
+  , takeExtension
+  , replaceExtension
+  , dropExtension
+  , addExtension
+  , hasExtension
+  , (<.>)
+  , splitExtensions
+  , dropExtensions
+  , takeExtensions
+  , splitExtension
+
+  -- * Filename\/directory functions
+  , splitFileName
+  , takeFileName
+  , replaceFileName
+  , dropFileName
+  , takeBaseName
+  , replaceBaseName
+  , takeDirectory
+  , replaceDirectory
+  , combine
+  , (</>)
+  , splitPath
+  , joinPath
+  , splitDirectories
+  , takeAllParents
+
+  -- * Drive functions
+  , splitDrive
+  , joinDrive
+  , takeDrive
+  , hasDrive
+  , dropDrive
+  , isDrive
+
+  -- * Trailing slash functions
+  , hasTrailingPathSeparator
+  , addTrailingPathSeparator
+  , dropTrailingPathSeparator
+
+  -- * File name manipulations
+  , normalise
+  , makeRelative
+  , equalFilePath
+  , isRelative
+  , isAbsolute
+  , isValid
+  , makeValid
+  , isFileName
+  , hasParentDir
+  )
+where
+
+import AFP.AbstractFilePath.Internal
+    ( afp
+    , bsToAFP
+    , fromAbstractFilePath
+    , fromAbstractFilePathIO
+    , toAbstractFilePath
+    , toAbstractFilePathIO
+    , unpackAFP
+    , packAFP
+    )
+import AFP.AbstractFilePath.Internal.Types
+    ( AbstractFilePath, PosixFilePath, WindowsFilePath )
+import AFP.OsString
+import AFP.OsString.Internal ( fromChar )
+import AFP.OsString.Internal.Types
+
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+import qualified AFP.AbstractFilePath.Windows as AFP
+#else
+import qualified AFP.AbstractFilePath.Posix as AFP
+#endif
+
+import Data.Bifunctor
+    ( bimap, first )
+
+
+------------------------
+-- Separator predicates
+
+
+-- | Ideal path separator character
+pathSeparator :: OsWord
+pathSeparator = OsWord AFP.pathSeparator
```

Comment:
Here (and in many other places), we could use `coerce`. Will that improve anything?